compiler/rustc_middle/src/query/mod.rs

```diff
@@ -2270,7 +2270,7 @@ rustc_queries! {
         desc { "check whether the item has a `where Self: Sized` bound" }
     }
 
-    query cross_crate_inlinable(def_id: DefId) -> bool {
+    query cross_crate_inlinable(instance: ty::InstanceKind<'tcx>) -> bool {
```

Comment:
It's probably best to split this into two queries -- something that takes `InstanceKind`, and something that works on raw `DefId`. The latter can be `separate_provide_extern`, maybe have some name like `*_raw` or `*_item`, and it'll continue to encode the mapping of `DefId -> bool` in the rmeta.

Then the `InstanceKind` one doesn't need to be `separate_provide_extern`, and instead can implement all the non `InstanceKind::Item` shim-specific behavior, which I assume does *not* have to be crate-specific. And honestly that one might not even need to be a query -- maybe just a regular tcx function or perhaps a hook so we can put the impl in `rustc_mir_transform`.

Does that idea make sense?

Reply:
Yeah. Such a query wrapper already exists, except it's spread across the monomorphization/reachability code. I was hoping this would be a separable change, but you're right that it probably isn't. I'm folding this approach into my other work.